File: packages/companion/swift/Sources/BridgethingCompanion/NluRejection.swift
import Foundation

public struct NluRejectionPolicy: Sendable, Equatable {
    public var inDomainThreshold: Double
    public var clarifyMargin: Double
    public var maxAlternates: Int

    public init(inDomainThreshold: Double = 0.5, clarifyMargin: Double = 0.15, maxAlternates: Int = 2) {
        self.inDomainThreshold = inDomainThreshold
        self.clarifyMargin = clarifyMargin
        self.maxAlternates = maxAlternates
    }
}

public enum NluRejectionOutcome: Sendable, Equatable {
    case accept(intent: String)
    case noIntent
    case clarify(alternates: [String])
}

public enum NluRejection {
    public enum RejectionError: Error, CustomStringConvertible {
        case headMismatch(logits: Int, catalog: Int)

        public var description: String {
            switch self {
            case let .headMismatch(logits, catalog):
                return "intent head emits \(logits) logits but the catalog has \(catalog) names"
            }
        }
    }

    public static func evaluate(
        _ output: NluInferenceOutput,
        policy: NluRejectionPolicy = .init()
    ) throws -> NluRejectionOutcome {
        let names = NluIntentCatalog.surfaceNames
        guard output.intentLogits.count == names.count else {
            throw RejectionError.headMismatch(logits: output.intentLogits.count, catalog: names.count)
        }

        guard sigmoid(output.inDomainLogit) >= policy.inDomainThreshold else { return .noIntent }

        let probabilities = softmax(output.intentLogits)
        let ranked = probabilities.enumerated()
            .sorted { $0.element > $1.element }
            .map { (name: names[$0.offset], probability: $0.element) }

        guard let top = ranked.first else { return .noIntent }
        guard let runnerUp = ranked.dropFirst().first else { return .accept(intent: top.name) }

        if top.probability - runnerUp.probability < policy.clarifyMargin {
            let alternates = ranked.prefix(max(policy.maxAlternates, 0)).map(\.name)
            return .clarify(alternates: Array(alternates))
        }
        return .accept(intent: top.name)
    }

    static func sigmoid(_ x: Double) -> Double {
        1 / (1 + exp(-x))
    }

    static func softmax(_ logits: [Double]) -> [Double] {
        guard let peak = logits.max() else { return [] }
        let exps = logits.map { exp($0 - peak) }
        let total = exps.reduce(0, +)
        guard total > 0 else { return Array(repeating: 0, count: logits.count) }
        return exps.map { $0 / total }
    }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/NluRejection.kt
package com.bridgething.companion

import kotlin.math.exp

sealed interface NluRejectionOutcome {
    data class Accept(val intent: String) : NluRejectionOutcome
    data object NoIntent : NluRejectionOutcome
    data class Clarify(val alternates: List<String>) : NluRejectionOutcome
}

object NluRejection {
    class HeadMismatch(logits: Int, catalog: Int) :
        Exception("intent head emits $logits logits but the catalog has $catalog names")

    fun evaluate(
        output: NluInferenceOutput,
        policy: NluRejectionPolicy = NluRejectionPolicy(),
    ): NluRejectionOutcome {
        val names = NluIntentCatalog.surfaceNames
        if (output.intentLogits.size != names.size) {
            throw HeadMismatch(output.intentLogits.size, names.size)
        }

        if (sigmoid(output.inDomainLogit) < policy.inDomainThreshold) return NluRejectionOutcome.NoIntent

        val ranked = softmax(output.intentLogits)
            .mapIndexed { index, probability -> names[index] to probability }
            .sortedByDescending { it.second }

        val top = ranked.firstOrNull() ?: return NluRejectionOutcome.NoIntent
        val runnerUp = ranked.getOrNull(1) ?: return NluRejectionOutcome.Accept(top.first)

        if (top.second - runnerUp.second < policy.clarifyMargin) {
            return NluRejectionOutcome.Clarify(ranked.take(maxOf(policy.maxAlternates, 0)).map { it.first })
        }
        return NluRejectionOutcome.Accept(top.first)
    }

    fun sigmoid(x: Double): Double = 1 / (1 + exp(-x))

    fun softmax(logits: List<Double>): List<Double> {
        val peak = logits.maxOrNull() ?: return emptyList()
        val exps = logits.map { exp(it - peak) }
        val total = exps.sum()
        if (total <= 0) return List(logits.size) { 0.0 }
        return exps.map { it / total }
    }
}
